proposals/accepted/50-ide-test-integration.md

```diff
@@ -0,0 +1,219 @@
+
+In summary, take each function signature and detect whether said function
+is runnable as a test based on whether its type is a member of the
+proposed typeclass `IsTest`. Testing frameworks will each provide their
+own `IsTest` instances for their common testing types.
+
+I have chosen the option in this proposal for the following reasons:
+ 1. Such an approach would run Haskell code in GHCi, which is slower, and
+ would therefore scale poorly for production codebases.
+ 2. Such an approach requires integration with GHC's API, which the author
+ is unfamiliar with and therefore cannot be confident in their ability to
+ maintain the integration library.
+ 3. The author cannot see how to acquire the test tree that is common to
+ IDE test integration, since it requires inspecting the typechecking artifacts
+ of all files in the test suite.
+ 4. Sometimes testing frameworks will provide conflicting `IsTest` instances for the
+ the same type, which is a problem when using multiple testing frameworks in the
+ same codebase.
+
+The typeclass proposal seems incredibly useful however for the typical `Run Program`
+command now provided, or as an additional tool for users. It is certainly a good
+idea for a future project.
+
+### Regex-based test discovery
+
+While regex discovery is a good and quick solution implementation wise, it does
+not scale to many real-world usecases. For example, many codebases will have helpers
+such as:
+
+```haskell
+testDecodeJsonFromFile fileName = testCase fileName $ readJsonFromFile fileName
+```
+
+Since the fileName value is not known until runtime, regex-based approaches miss
+such cases, which is a non-starter for an ecosystem-wide tool.
+
+The proposal put forth by the author will also have to handle this use-case, by
+providing support in testing frameworks for users to define their own combinators
+that declare their location on use site. But this problem is a purely technical
+one with many solutions, such as the following pseudo-code:
+
+```haskell
+testDecodeJsonFromFile :: HasCallStack => TestTree
+testDecodeJsonFromFile = Tasty.customTestCase callstack fileName $ readJsonFromFile fileName
+```
+
+## Future Work
+
+This proposal is purposefully minimal in scope, for a number
+of reasons:
+ 1. In order for testing frameworks to buy in, we will need a
+ minimally-controversial interface.
+ 2. Support for more features should be done incrementally, since
+ the design space here is rather large, and we will not know users'
+ needs until after integration is alright out in the real world.
+
+Here are some features that have been neglected in this proposal:
+ 1. Distinction between "error" and "failure" when a test fails.
+ 2. Support for multiple filters when running a test.
+ 3. Support for options when running a test, such as: `--accept` when
+ running golden tests, `-jN` to run tests in parallel, and options
+ related to property-based tests.
+ 4. Support for localized options when running a test.
```

Comment:
What are localized options? In my head it would be something like replacing `--accept` with `--godtage` to make the Danish localization, but this doesn't seem like a very plausible interpretation of what you're writing.